src/invariant/BucketListIsConsistentWithDatabase.cpp

```diff
@@ -258,6 +259,96 @@ BucketListIsConsistentWithDatabase::checkEntireBucketlist()
     }
 }
 
+std::string
+BucketListIsConsistentWithDatabase::checkAfterAssumeState(uint32_t newestLedger)
+{
+    // If BucketListDB is disabled, we've already enforced the invariant on a
+    // per-Bucket level
+    if (!mApp.getConfig().isUsingBucketListDB())
+    {
+        return {};
+    }
+
+    EntryCounts counts;
+    LedgerKeySet seenKeys;
+
+    auto perBucketCheck = [&](auto bucket, auto& ltx) {
+        for (BucketInputIterator iter(bucket); iter; ++iter)
+        {
+            auto const& e = *iter;
+
+            if (e.type() == LIVEENTRY || e.type() == INITENTRY)
+            {
+                if (e.liveEntry().data.type() != OFFER)
+                {
+                    continue;
+                }
+
+                // If this is the newest version of the key in the BucketList,
+                // check against the db
+                auto key = LedgerEntryKey(e.liveEntry());
+                auto [_, newKey] = seenKeys.emplace(key);
+                if (newKey)
+                {
+                    counts.countLiveEntry(e.liveEntry());
+
+                    auto s = checkAgainstDatabase(ltx, e.liveEntry());
+                    if (!s.empty())
+                    {
+                        return s;
+                    }
+                }
+            }
+            else if (e.type() == DEADENTRY)
+            {
+                if (e.deadEntry().type() != OFFER)
+                {
+                    continue;
+                }
+
+                // If this is the newest version of the key in the BucketList,
+                // check against the db
+                auto [_, newKey] = seenKeys.emplace(e.deadEntry());
+                if (newKey)
+                {
+                    auto s = checkAgainstDatabase(ltx, e.deadEntry());
+                    if (!s.empty())
+                    {
+                        return s;
+                    }
+                }
+            }
+        }
+
+        return std::string{};
+    };
+
+    {
+        LedgerTxn ltx(mApp.getLedgerTxnRoot());
+        auto& bl = mApp.getBucketManager().getBucketList();
+
+        for (uint32_t i = 0; i < BucketList::kNumLevels; ++i)
+        {
+            auto const& level = bl.getLevel(i);
+            for (auto const& bucket : {level.getCurr(), level.getSnap()})
+            {
+                auto s = perBucketCheck(bucket, ltx);
+                if (!s.empty())
+                {
+                    return s;
+                }
+            }
+        }
+    }
+
+    auto range =
+        LedgerRange::inclusive(LedgerManager::GENESIS_LEDGER_SEQ, newestLedger);
+
+    // SQL only stores offers when BucketListDB is enabled
+    return counts.checkDbEntryCounts(
```

Comment:
I'm not sure if that matters much, but could that check go to `checkOnBucketApply`? It seems like that would be a bit more consistent.

Reply:
No, this check makes sure the number of entries in SQL DB matches the number of entries in the BucketList. This is not possible to check on a per-bucket level when applying buckets in-order. For example, suppose an offer `DEADENTRY` exists on level 0, and a `LIVEENTRY` of the same key exists on level 1. The database correctly does not contain the entry such that the offer count is 0. When performing the check on the level 1 bucket, the invariant would fail, as based on that individual bucket, the offer count should be 1 instead of 0.

Essentially we can only do individual Bucket level checks in `checkOnBucketApply`, whereas `checkAfterAssumeState` performs BucketList level checks.

Reply:
Thanks for the explanation.